ouroboros-network/test/Test/Ouroboros/Network/TxSubmission.hs

```diff
@@ -261,26 +267,103 @@ instance Arbitrary a => Arbitrary (LargeNonEmptyList a) where
                   -> NonEmptyList (Tx Int)
                   -> Maybe Delay
-                  -> Bool
+                  -> Property
 prop_txSubmission (Positive maxUnacked) (NonEmpty outboundTxs) delay =
     let mbDelayTime = getDelay <$> delay
-        inboundTxs =
-          runSimOrThrow $ do
+        tr = (runSimTrace $ do
             controlMessageVar <- newTVarIO Continue
             _ <-
               async $ do
                 threadDelay
                   (fromMaybe 1 mbDelayTime
-                    * (realToFrac (length outboundTxs `div` 4)))
+                    * realToFrac (length outboundTxs `div` 4))
                 atomically (writeTVar controlMessageVar Terminate)
             txSubmissionSimulation
               maxUnacked outboundTxs
               (readTVar controlMessageVar)
               mbDelayTime mbDelayTime
-              
-    in inboundTxs == take (length inboundTxs) outboundTxs
+            ) in
+    ioProperty $ do
+        tr' <- evaluateTrace tr
+        case tr' of
+             SimException e trace -> do
+                return $ counterexample (intercalate "\n" $ show e : trace) False
+             SimDeadLock trace -> do
+                 return $ counterexample (intercalate "\n" $ "Deadlock" : trace) False
+             SimReturn (inmp, outmp) _trace -> do
+                 -- printf "Log: %s\n" (intercalate "\n" _trace)
+                 let outUniqueTxIds = nubBy (on (==) getTxId) outmp
+                     outValidTxs    = filter getTxValid outmp
+                 case (length outUniqueTxIds == length outmp, length outValidTxs == length outmp) of
+                      (True, True) ->
+                          -- If we are presented with a stream of unique txids for valid
+                          -- transactions the inbound transactions should match the outbound
+                          -- transactions exactly.
+                          return $ inmp === take (length inmp) outValidTxs
+                      (True, False) ->
+                          -- If we are presented with a stream of unique txids then we should have
+                          -- fetched all valid transactions.
+                          return $ inmp === take (length inmp) outValidTxs
+                      (False, True) ->
+                          -- If we are presented with a stream of valid txids then we should have
+                          -- fetched some version of those transactions.
+                          return $ map getTxId inmp === take (length inmp) (map getTxId $
+                              filter getTxValid outUniqueTxIds)
+                      (False, False)
+                           -- If we are presented with a stream of valid and invalid Txs with
+                           -- duplicate txids we're content with completing the protocol
+                           -- without error.
+                           -> return $ property True
+
+
+-- TODO: Belongs in iosim.
```

Comment:
I have no interest in including `go` directly into my property. I prefer to keep it in a form that makes it easy to cut-and-paste into every future test case I write.

Reply:
With that feedback, I'll refine my suggestion to define a combinator like this one. The main refinement is that it doesn't necessitate the `[String]` space-leak and so seems like a good candidate for inclusion in `io-sim`.

```
foldSayResultTrace ::
     (Time -> ThreadId -> String -> a -> a)   -- ^ handle an 'EventSay'
  -> (a -> Either Failure b -> c)  -- ^ handle end of trace
  -> (a -> SomeException -> c)  -- ^ handle pure exception thrown by definition of trace
  -> a
  -> Trace b
  -> c
```

Main point: Whatever the details end up being, I agree that a special combinator focusing only on `say` and termination events would be helpful. It could be another `TraceSay` data type with its own fold function or just the Church encoding as above, no personal preference. For example, I'm mainly envisioning `a ~ c ~ Test.QuickCheck.Property`.